Update provide_bucket_name() decorator to handle new conn_type (#28706)
mistercrunch/Airflow: docs/apache-airflow-providers-amazon/connections/aws.rst

    Specify the extra parameters (as json dictionary) that can be used in AWS
    connection. All parameters are optional.

    * ``s3_bucket_name``: Name of the S3 bucket that the hook should reference

> 

Reviewer: I'd like the idea for move this setting in separate extra rather than use `schema`
However I don't think service level parameter should stored in top level of AWS Connection Extra.

The reason is simple - `boto3` provide access to 300+ AWS Services, and if we store it in top level, then we have a huge chance to have a mess in the future like we have now mess with credentials and assuming roles

```python
Python 3.9.10 (main, Feb 25 2022, 16:54:01) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.7.0 -- An enhanced Interactive Python. Type '?' for help.
PyDev console: using IPython 8.7.0
Python 3.9.10 (main, Feb 25 2022, 16:54:01) 
[Clang 13.0.0 (clang-1300.0.29.30)] on darwin

import boto3, botocore

botocore.__version__
Out[3]: '1.29.43'

boto3.__version__
Out[4]: '1.26.43'

session = boto3.session.Session(region_name="us-east-1")
len(session.get_available_services())
Out[6]: 336
```

I'd suggest to move it in separate parameter (dictionary) which store per service level, e.g.

```json
{
  "service_config": {
    "s3": {
      "bucket_name": "awesome-bucket"
    },
    "sts": {
      "endpoint_url": "https://example.org"
    },
    "emr": {
      "job_flow_overrides": {"Name": "PiCalc", "ReleaseLabel": "emr-6.7.0"},
      "endpoint_url": "https://emr.example.org"
    }
}

```

This changes also might help with this PR: https://github.com/apache/airflow/pull/28545 (cc: @confusedpublic)
And in long term allow us discontinue [Amazon Elastic MapReduce (EMR) Connection](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/emr.html#amazon-elastic-mapreduce-emr-connection)

Author: That sounds good. I have updated airflow/providers/amazon/aws/hooks/s3.py to reflect the requested changes. From these changes, the user would be able to specify the S3 bucket name in the "extra" parameter when creating an aws connection, by entering the following json:
```
{
  "service_config": {
    "s3": {
      "bucket_name": "[bucket name]"
    },
}
```
Let me know if this is what you had in mind, or if there are any other changes that are needed. 
Also, I was not sure how to update the documentation to reflect this option of indicating S3 Bucket name. Would I just specify in `docs/apache-airflow-providers-amazon/connections/aws.rst` that the user should enter `"service_config.s3.bucket_name"` in `Extra's` in order to specify the bucket name, or is there a better way to indicate this?

    Specify the extra parameters (as json dictionary) that can be used in AWS
    connection. All parameters are optional.

>     * ``service_config``: json used for specifying information about different AWS services, such as S3 or STS.

Reviewer 2: ```suggestion
    * ``service_config``: json used to specify configuration/parameters for different AWS services, such as S3 or STS.
```

      "assume_role_kwargs": { "something":"something" }
    }

> 5. Using **service_config** to specify information about services such as S3, STS, and EMR

Reviewer 2: ```suggestion
5. Using **service_config** to specify configuration for services such as S3, STS, and EMR
```
